File: scripts/src/generate-migration.ts
#!/bin/node
import fs from 'node:fs';
import { spawnSync } from 'child_process';
import { argv } from 'node:process';


/**
 * Add EF core migration file, edited to be compatible with PublishTrimmed.
 */
const addMigrationTrimmedCompatible = async (filename: string) => {
    console.log(`Create migration with filename=${filename}`);

    if (!filename) {
        throw new Error('no filename given');
    }

    const proc = spawnSync(
        "dotnet",
        `ef migrations add ${filename} --json`.split(' '),
        {
            cwd: '../PKVault.Backend',
        }
    );

    const error = proc.stderr.toString();
    const output = proc.stdout.toString();

    console.log(`stderr: ${error}`);
    console.log(`stdout: ${output}`);
    console.log();

    if (proc.error) {
        throw proc.error;
    }

    if (error) {
        throw new Error(error);
    }

    const jsonTxt = output.split('\n')
        .filter(line => line.trim().length > 0)
        .slice(-5)
        .join('\n');

    const json: { [ key in string ]?: string } = JSON.parse(jsonTxt);
    const migrationFilePath = json.migrationFile;
    if (!migrationFilePath) {
        throw new Error('no migrationFilePath found');
    }

    const migrationContent = fs.readFileSync(migrationFilePath, 'utf8');

    const migrationContentFixed = migrationContent.split('\n')
        .map(withName)
        .join('\n');

    fs.writeFileSync(migrationFilePath, migrationContentFixed, 'utf8');

    console.log("migrationContentFixed:", migrationContentFixed);
};

const withName = (line: string): string => {
    // Regex catching:
    // - column name ("ID" etc)
    // - parameters (type: "...", ...)
    const pattern = /^\s*(\w+)\s*=\s*table\.Column<.+>\(([^)]*)\),?\s*$/;

    const match = pattern.exec(line);

    if (!match) {
        return line;
    }

    const columnName = match[ 1 ];   // ex: "Id"
    const parameters = match[ 2 ];   // ex: "type: "...", nullable: false"

    if (
        !columnName.trim()
        || !parameters.trim()
        || parameters.includes('name:')
    ) {
        return line;
    }

    if (!parameters.includes("type:")) {
        throw new Error(`Missing 'type:' with line:\n${line}`);
    }

    return line.replace("type:", `name: "${columnName}", type:`);
}

addMigrationTrimmedCompatible(argv[ 2 ] ?? '')

File: scripts/update-pkhex/GenerateMigration.cs

using System.Diagnostics;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using Serilog;

public partial class GenerateMigration
{
    public static async Task GenerateTrimmedCompatibleMigration(string filename)
    {
        Log.Information($"Create migration with filename={filename}");

        ArgumentException.ThrowIfNullOrWhiteSpace(filename);

        var psi = new ProcessStartInfo
        {
            FileName = "dotnet",
            Arguments = $"ef migrations add {filename} --json",
            WorkingDirectory = "../PKVault.Core",
            RedirectStandardError = true,
            RedirectStandardOutput = true,
        };

        using var proc = Process.Start(psi)!;
        await proc.WaitForExitAsync();

        var error = await proc.StandardError.ReadToEndAsync();
        var output = await proc.StandardOutput.ReadToEndAsync();

        Log.Debug($"stderr: {error}");
        Log.Debug($"stdout: {output}");
        Console.WriteLine();

        if (proc.ExitCode > 0) {
            Environment.Exit(proc.ExitCode);
        }

        var jsonTxt = string.Join('\n', output.Split('\n')
            .Where(line => line.Trim().Length > 0)
            .TakeLast(5)
        );

        var json = JsonSerializer.Deserialize<Dictionary<string, string>>(jsonTxt)!;

        if (!json.TryGetValue("migrationFile", out var migrationFilePath))
            throw new Exception("no migrationFilePath found");

        var migrationContent = await File.ReadAllTextAsync(migrationFilePath, Encoding.UTF8);

        var migrationContentFixed = string.Join('\n',
            migrationContent.Split('\n')
                .Select(WithName)
        );

        await File.WriteAllTextAsync(migrationFilePath, migrationContentFixed, Encoding.UTF8);

        Log.Debug($"migrationContentFixed: {migrationContentFixed}");
    }
    
    public static string WithName(string line)
    {
        var match = Pattern().Match(line);

        if (match.Length == 0) {
            return line;
        }

        var columnName = match.Groups[ 1 ].Value.Trim();   // ex: "Id"
        var parameters = match.Groups[ 2 ].Value.Trim();   // ex: "type: "...", nullable: false"

        if (
            columnName.Length == 0
            || parameters.Length == 0
            || parameters.Contains("name:")
        ) {
            return line;
        }

        if (!parameters.Contains("type:")) {
            throw new Exception($"Missing 'type:' with line:\n{line}");
        }

        return line.Replace("type:", $"name: \"{columnName}\", type:");
    }

    // Regex catching:
    // - column name ("ID" etc)
    // - parameters (type: "...", ...)
    [GeneratedRegex(@"^\s*(\w+)\s*=\s*table\.Column<.+>\(([^)]*)\),?\s*$")]
    private static partial Regex Pattern();
}
